### PR TITLE
[Minor] Make LRUCache use LRU strategy

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Make LRUCache use LRU strategy.  

Make limiting strategy configurable for `LimitedMap` and construct `LimitedSet` by wrapping a `LimitedMap`.

### CODE REVIEW

File: util/src/main/java/tech/pegasys/artemis/util/collections/LimitedMap.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+   * @param maxSize The maximum number of elements to keep in the map.
+   * @param mode A mode that determines which element is evicted when the map exceeds its max size.
+   * @param <K> The key type of the map.
+   * @param <V> The value type of the map.
+   * @return A thread-safe map that will evict elements when the max size is exceeded.
+   */
+  public static final <K, V> Map<K, V> create(
+      final int initialCapacity, final int maxSize, final LimitStrategy mode) {
+    final boolean useAccessOrder = mode.equals(LimitStrategy.DROP_LEAST_RECENTLY_ACCESSED);
+    return Collections.synchronizedMap(
```
Comment (ajsutton) [inline util/src/main/java/tech/pegasys/artemis/util/collections/LimitedMap.java]: I'm not sure we should move the `synchronizedMap` into here.  I think it makes sense to potentially have instances of `LimitedMap` that are only accessed by a single thread or where there are already other locking mechanisms in place.  It also follows the typical pattern of collection classes not being thread-safe unless they have `Concurrent` in the name.

Comment (ajsutton) [reply util/src/main/java/tech/pegasys/artemis/util/collections/LimitedMap.java]: I see that we'd previously had the `synchronizedSet` applied automatically to `LimitedSet` so if we did move this we'd have to ensure we added that to each place `LimitedSet` is constructed.

**Review by ajsutton — APPROVED** [review]
LGTM. I would lean towards leaving the synchronization a responsibility of the caller but it doesn't make much difference at the moment so not a big deal.

@Nashatyrev Would be good to get your input on this as well - was there any particular reason to favour least recently added vs least recently accessed?

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java [hunk 2]
```diff
@@ -42,12 +43,12 @@ public LRUCache(int capacity) {
 
   private LRUCache(int capacity, Map<K, V> initialCachedContent) {
     this.maxCapacity = capacity;
-    LimitedHashMap<K, V> cacheMap = new LimitedHashMap<>(maxCapacity);
+    Map<K, V> cacheMap = LimitedMap.create(maxCapacity, LimitStrategy.DROP_LEAST_RECENTLY_ACCESSED);
     // copy safely, initialCachedContent is always a SynchronizedMap instance
     synchronized (initialCachedContent) {
```
Comment (Nashatyrev) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:49]: I agree with @ajsutton that synchronizing a collection would better to perform on a client side. This place is a good illustration: we need to make sure that the collection is `synchronized` here, but `LimitedMap.create` doesn't imply returning synchronized collection. 
Alternative option would be to have two factory variants like: 
- `LimitedMap.create`
- `LimitedMap.createSynchronized`

**Comment by Nashatyrev** [comment]
> @Nashatyrev Would be good to get your input on this as well - was there any particular reason to favour least recently added vs least recently accessed?

I think the only reason was that I just didn't know about that access/add `mode` flag :)
Otherwise recently accessed mode should fit better for any cache IMHO

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java (hunk 2 above)
Comment (mbaxter) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java:49]: Good points!  I split up the helpers so we now have:
- `LimitedMap.create`
- `ConcurrentLimitedMap.create`
- `LimitedSet.create`
- `ConcurrentLimitedSet.create`

**Pull request merged**